src/lib/support/PersistedCounter.h

```diff
@@ -104,16 +105,50 @@ class PersistedCounter : public MonotonicallyIncreasingCounter<T>
         }
 #endif
 
-        ReturnErrorOnFailure(PersistNextEpochStart(startValue + aEpoch));
+        ReturnErrorOnFailure(PersistNextEpochStart(static_cast<T>(startValue + aEpoch)));
 
         // This will set the starting value, after which we're ready.
         return MonotonicallyIncreasingCounter<T>::Init(startValue);
     }
 
     /**
-     *  @brief
-     *  Increment the counter and write to persisted storage if we've completed
-     *  the current epoch.
+     *  @brief Increment the counter by N and write to persisted storage if we've completed the current epoch.
+     *
+     *  @param value value of N
+     *
+     *  @return Any error returned by a write to persisted storage.
+     */
+    CHIP_ERROR AdvanceBy(T value) override
+    {
+        VerifyOrReturnError(mStorage != nullptr, CHIP_ERROR_INCORRECT_STATE);
+        VerifyOrReturnError(mKey.IsInitialized(), CHIP_ERROR_INCORRECT_STATE);
+
+        // If value is 0, we do not need to do anything
+        VerifyOrReturnError(value > 0, CHIP_NO_ERROR);
+
+        // We should update the persisted epoch value if :
+        // 1- Sum of the current counter and value is greater or equal to the mNextEpoch.
+        //    This is the standard operating case.
+        // 2- Increasing the current counter by value would cause a roll over. This would cause the current value to be < to the
+        //    mNextEpoch so we force an update.
+        bool shouldDoEpochUpdate = ((MonotonicallyIncreasingCounter<T>::GetValue() + value) >= mNextEpoch) ||
+            (MonotonicallyIncreasingCounter<T>::GetValue() > std::numeric_limits<T>::max() - value);
+
+        ReturnErrorOnFailure(MonotonicallyIncreasingCounter<T>::AdvanceBy(value));
+
+        if (shouldDoEpochUpdate)
+        {
+            // Since AdvanceBy allows the counter to be increase by an arbitrary value, it is possible that the new counter value is
+            // greater than mNextEpoch + mEpoch. As such, we want the next Epoch value to be calculate from the new current
+            // value.
```

Comment:
```suggestion
            // Since AdvanceBy allows the counter to be increased by an arbitrary value, it is possible that the new counter value is
            // greater than mNextEpoch + mEpoch. As such, we want the next Epoch value to be calculated from the new current
            // value.
```